### PR TITLE
[SPARK-8301][SQL] Improve UTF8String substring/startsWith/endsWith/contains performance

### PR DESCRIPTION
**Pull request by tarekbecker**
Jira: https://issues.apache.org/jira/browse/SPARK-8301

Added the private method startsWith(prefix, offset) to implement startsWith, endsWith and contains without copying the array

I hope that the component SQL is still correct. I copied it from the Jira ticket.

### CODE REVIEW

**Comment by rxin** [comment]
Jenkins, test this please.

**Comment by rxin** [comment]
cc @davies for review.

**Comment by davies** [comment]
@tarekauel Thanks for working on this, could you have some micro benchmark to show the performance difference? It's good to see the improvements by numbers.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 1]
```diff
@@ -131,24 +131,31 @@ public boolean contains(final UTF8String substring) {
...
-      // TODO: Avoid copying.
-      if (bytes[i] == b[0] && Arrays.equals(Arrays.copyOfRange(bytes, i, i + b.length), b)) {
+      if (bytes[i] == b[0] && startsWith(substring, i)) {
         return true;
       }
     }
     return false;
   }
 
+  private boolean startsWith(final UTF8String prefix, int offset) {
```
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Nit: If we use `startsWith(final byte[] prefix, int offset)`, it will may save some calls of `s.getBytes()`. (not sure how is the difference).

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 2]
```diff
@@ -131,24 +131,31 @@ public boolean contains(final UTF8String substring) {
...
+    byte[] b = prefix.getBytes();
+    if (b.length + offset > bytes.length || offset < 0) {
+      return false;
+    }
+    int i = 0;
+    while (i < b.length && b[i] == bytes[i + offset]) {
```
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: As @rxin suggested, we could use unsafe API to compare 8 bytes at once. But I think in most cases the suffix will be short and this while loop will terminate very fast, it may not worth that complication.

The while loop is good to me.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 3]
```diff
@@ -131,24 +131,31 @@ public boolean contains(final UTF8String substring) {
...
+      i++;
+    }
+    return i == b.length;
+  }
+
   public boolean startsWith(final UTF8String prefix) {
-    final byte[] b = prefix.getBytes();
-    // TODO: Avoid copying.
-    return b.length <= bytes.length && Arrays.equals(Arrays.copyOfRange(bytes, 0, b.length), b);
+    return startsWith(prefix, 0);
```
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Right now, UTF8String is only used by catalyst, can it will check prefix is null or not. Since this is a public method, it's still good to check prefix is null or not.

**Comment by tarekauel** [comment]
@davies thanks for your feedback.
I added a null check and test cases for null. There are other public methods without null checks. Should I add a null check to them as well?

I did a micro benchmark for starsWith. I created objects for the UTF8String and called the test 1,000,000,000 times.

duration starts with old: 73,586 ms
duration starts with new: 23,246 ms (just 31.6%)

**Comment by davies** [comment]
@tarekauel LGTM. It will be better if you could also add null check for others, thanks for the numbers, good to know the improvements!

**Comment by tarekauel** [comment]
@davies I added the null checks. If someone compares a UTF8String and null, the null value will come first and the string afterwards. For set I decided to initialise an empty byte array, in order to avoid a crash later.

**Comment by davies** [comment]
@tarekauel I think we could mark `set` as protected(only used in a few places, include codegen), make sure that the parameters is not null. Could you do it in this PR?

**Comment by tarekauel** [comment]
@davies I have changed the access modifier and I changed the annotation of bytes to Nonnull from Nullable. I added all non null checks, didn't I?

**Comment by davies** [comment]
@tarekauel Thanks for the change, you need to update all the usage of (new UTF8String().set) to UTF8String.fromString/Bytes, or it can not compile or fail the tests.

**Comment by tarekauel** [comment]
@davies done. Intellij found only one reference. Can you trigger the build process again?

**Comment by davies** [comment]
you can `git grep UTF8String` | grep set

**Comment by tarekauel** [comment]
There's just one other file ```sql/core/src/main/scala/org/apache/spark/sql/columnar/ColumnType.scala```this uses already ```UTF8.fromBytes(...)```

**Comment by davies** [comment]
Sorry, it should be `git grep stringType | grep set`

**Comment by tarekauel** [comment]
Thanks!

**Comment by tarekauel** [comment]
Can someone trigger Jenkins?

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by tarekauel** [comment]
@davies I fixed it. The change was correct, wasn't it? Can someone start Jenkins again?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 4]
```diff
@@ -437,17 +437,17 @@ case class Cast(child: Expression, dataType: DataType) extends UnaryExpression w
 
       case (BinaryType, StringType) =>
         defineCodeGen (ctx, ev, c =>
-          s"new ${ctx.stringType}().set($c)")
+          s"new ${ctx.stringType}().fromBytes($c)")
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: This should be `${ctx.stringType}.fromBytes($c)`, also others

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 5]
```diff
@@ -55,13 +55,17 @@ public static UTF8String fromString(String str) {
...
-  public UTF8String set(final String str) {
-    try {
-      bytes = str.getBytes("utf-8");
-    } catch (UnsupportedEncodingException e) {
-      // Turn the exception into unchecked so we can find out about it at runtime, but
-      // don't need to add lots of boilerplate code everywhere.
-      PlatformDependent.throwException(e);
+  protected UTF8String set(final String str) {
+    if (str == null) {
+      bytes = new byte[0];
```
Comment (chenghao-intel) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: This's probably not right, as a null string will be converted into empty string "". 
(new UTF8String().set(null).toString() == "") ? Can we revert this change?

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 6]
```diff
@@ -69,8 +73,8 @@ public UTF8String set(final String str) {
   /**
    * Updates the UTF8String with byte[], which should be encoded in UTF-8.
    */
-  public UTF8String set(final byte[] bytes) {
-    this.bytes = bytes;
+  protected UTF8String set(final byte[] bytes) {
+    this.bytes = (bytes != null) ? bytes : new byte[0];
```
Comment (chenghao-intel) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Will it be better if we throw exception if null passed in? Or leave it as it was.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 5 above)
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: It's safe to remove this check, because `set` is protected now, we can make sure that `str` is not null.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 7]
```diff
@@ -125,30 +129,37 @@ public UTF8String substring(final int start, final int until) {
...
-      // TODO: Avoid copying.
-      if (bytes[i] == b[0] && Arrays.equals(Arrays.copyOfRange(bytes, i, i + b.length), b)) {
+      if (bytes[i] == b[0] && startsWith(b, i)) {
         return true;
       }
     }
     return false;
   }
 
+  private boolean startsWith(final byte[] prefix, int offset) {
```
Comment (chenghao-intel) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: At the first sight, I thought the `offset` is for the start point of the `prefix`. However, if the `offset` represents the start pos of the current `UTF8String.bytes` in the comparison, probably we need to change the method name to `contains` or something, and switch the `offset` and `prefix`.

File: unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java [hunk 8]
```diff
@@ -46,6 +46,7 @@ public void basicTest() throws UnsupportedEncodingException {
 
   @Test
   public void contains() {
+    Assert.assertFalse(UTF8String.fromString("hello").contains(null));
```
Comment (chenghao-intel) [inline unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java]: I was trying the "a".contains(null) in java, and it throws exception like:
```
scala> "a".contains(null)
java.lang.NullPointerException
        at java.lang.String.contains(String.java:2076)
        at .<init>(<console>:8)
        at .<clinit>(<console>)
        at .<init>(<console>:7)
        at .<clinit>(<console>)
        at $print(<console>)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
```
Should we follow the same pattern for `UTF8String`?

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 9]
```diff
@@ -178,6 +189,7 @@ public UTF8String clone() {
 
   @Override
   public int compareTo(final UTF8String other) {
+    if (other == null) return 1;
```
Comment (chenghao-intel) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]:
```
scala> "a".compareTo(null)
java.lang.NullPointerException
        at java.lang.String.compareTo(String.java:1139)
        at .<init>(<console>:8)
        at .<clinit>(<console>)
        at .<init>(<console>:7)
        at .<clinit>(<console>)
        at $print(<console>)
```

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 7 above)
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: or `match`?

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 9 above)
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: I think we should support ordering a column with null in it, even Scala can't.

**Comment by chenghao-intel** [comment]
@tarekauel Glad to see the big improvement for the performance gain! I have some comments on the API change, e.g. the `null` value checking, comparison with `null`, probably we'd better not change its original behavior for `UTF8String`, seems behavior as `java.lang.String` does will be reasonable to me, what do you think?

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 7 above)
Comment (tarekauel) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: This is adopted from the original java String API. But yes it's true that ```offset``` depends on the bytes of the string and not on characters. Shall I rename it to ```match```? I could add a comment that ```offset`` depends on the bytes representation and not on characters

Comment (rxin) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: how about remaining it offsetInBytes?

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 5 above)
Comment (tarekauel) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: okay I am going to remove this

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 6 above)
Comment (tarekauel) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: If we don't want to accept `null`, I'd prefer to throw an exception. It's much easier to find a bug later, if the `set` throws an exception instead of any other method that runs into a null pointer exception. What do you think?

Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Same to my another comment, this `bytes` will not be `null`.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 9 above)
Comment (tarekauel) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: I think it should be able to sort columns with `null` values. The behavior should be compared to a database, shouldn't it?

Comment (chenghao-intel) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: The expression `RowOrdering` actually did the null testing already before call the `compareTo`, see https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/rows.scala#L218
So, we are sure the argument passed in will be non-null.

**Comment by tarekauel** [comment]
@chenghao-intel First of all thanks for all your comments and thoughts. I guess the `UTF8String` is only used in SPARK SQL, isn't it? Because of that I think the behavior should be similar to a database and not to JAVA. But I don't know if this should be achieved by modifying the SQL layer or this type implementation. 

What do you think?

**Comment by marmbrus** [comment]
+1, UTF8String should follow SQL semantics not JVM semantics.

**Comment by chenghao-intel** [comment]
I know the null checking stuff will make our code more safe, but we usually  don't do that in the API implementation, as it saves lots of redundant of code, that's almost the general conversion, otherwise we need to make a note in the scaladoc/javadoc. 

Expression framework (including the `UTF8String`) is a quite independent system to SQL, follow the general API design principles will be OK, and I believe it's the responsibility for SQLLayer to make the right calls.

**Comment by tarekauel** [comment]
Yeah, that's a good comment. I checked the String expressions and it does already the null check, see

https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala#L173-L182

So it won't be null and a valid result will be provided, if a value is null. I think the result is even better, because it returns null and not false which is way better.

**Comment by chenghao-intel** [comment]
Probably be `Nonnull`, this will give more strict check by FindBug. It will be great if you can run FindBug locally after the change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 10]
```diff
@@ -437,17 +437,17 @@ case class Cast(child: Expression, dataType: DataType) extends UnaryExpression w
 
       case (BinaryType, StringType) =>
         defineCodeGen (ctx, ev, c =>
-          s"new ${ctx.stringType}().set($c)")
+          s"${ctx.stringType}().fromBytes($c)")
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: No `()` here

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 11]
```diff
@@ -437,17 +437,17 @@ case class Cast(child: Expression, dataType: DataType) extends UnaryExpression w
...
       case (DateType, StringType) =>
         defineCodeGen(ctx, ev, c =>
-          s"""new ${ctx.stringType}().set(
+          s"""${ctx.stringType}().fromString(
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: No `()` here

**Comment by tarekauel** [comment]
All should be fine now. Can someone trigger Jenkins?

**Comment by davies** [comment]
LGTM, waiting for tests.
